### DISCUSSION TITLE
Nested parallelism with pytest-xdist and MPI/OpenMP

### DISCUSSION THREAD

**Discussion by gnikit** [pytest-dev/pytest-xdist#1141] [General]
I have a test suite containing _N_ tests, meant to run some long and computationally intensive programs, each test class spins up _M_ processes (the number of processes varies from class to class). I was wondering if there was a way to use xdist and some custom weights to load balance the work in the scheduler. Any thoughts?

## MWE

This is a dummy example using MPI.

```py
from subprocess import run


class TestParallelMyApp1:
    CORES = 4

    def test_main(self):
        run(["mpiexec", "-n", str(self.CORES), "python3", "-c", "print('App1')"])


class TestParallelMyApp2:
    CORES = 14

    def test_main(self):
        run(["mpiexec", "-n", str(self.CORES), "python3", "-c", "print('App2')"])


class TestParallelMyApp3:
    CORES = 10

    def test_main(self):
        run(["mpiexec", "-n", str(self.CORES), "python3", "-c", "print('App3')"])
```

### Desired Outcome

On a machine with 14 cores, running `pytest -n auto` would run:
- `TestParallelMyApp1` and `TestParallelMyApp3`
- and then `TestParallelMyApp2`

**Comment by gnikit** [comment]
Is there any idea on this from the dev team? It would be good to know if there are any solutions using `pytest-xdist`

**Reply by RonnyPfannschmidt** [reply]
As far as I'm aware none of us has mpi experience 

My last contact with mpi was about 17 years ago in a university course 

*edit *

And that involved the c variant

**Reply by gnikit** [reply]
Thanks for taking the time to reply @RonnyPfannschmidt, I appreciate it. 

It's not an MPI specific issue. I'm just using pytest to spin up a bunch of expensive processes (outside of the python runtime), where each one uses different cpu resources and the only thing I would like is for pytest-xdist to be aware that test_a is using 2 cores and test_b 4, so in a machine with 4 cores doing pytest-n 2 will not oversubscribe thus increasing the test suites runtime.

You can see that if you have dozens of such tests, calculating the ordering and grouping of the tests a-priori becomes big problem and oversubscribed jobs can cause substantial a slow down.

Basically any way to assign weights to xdist is what I'm looking for.

**Reply by RonnyPfannschmidt** [reply]
Currently there's no such mechanism 

We intend to add support for information passover to the sheduler 

However we also lack a sheduler tgats aware of such weights/ costs 

Its similar to loadscope but basically loadcost

**Reply by gnikit** [reply]
I think I might have managed to make a solution to this. I have not yet tested extensively, so I can't make any promises. If it works would you be keen for me to open a PR and merge it in?

**Reply by RonnyPfannschmidt** [reply]
im curious about the solution but without seeing it i certainly cant promise a merge

we already made the mistake of letting loadgroup modify nodeids - so i want to be very arefull about what debt/solution we bring in

**Reply by gnikit** [reply]
I've come up with a few equally problematic solutions, but I will try and work on it a bit today and see if I can get it to work. I'll post a MWE when I'm done. Cheers for the help @RonnyPfannschmidt

**Reply by gnikit** [reply]
Apologies @RonnyPfannschmidt forgot about this and posting my solution. I hacked something together that seems to work for us but it's  definitely suboptimal in many ways and hard to maintain. The correct thing to do would be to make an xdist scheduler class but I ran out of time. The hacky code looks something like this.

```python
import fcntl
import json
import os
import time
from pathlib import Path

import pytest


def pytest_addoption(parser):
    parser.addoption(
        "--weight-poll-interval",
        action="store",
        type=float,
        default=0.2,
        dest="weight_poll_interval",
        help="Seconds between retries when waiting for weighted capacity.",
    )
    parser.addoption(
        "--weight-timeout",
        action="store",
        type=float,
        default=0.0,
        dest="weight_timeout",
        help=(
            "Timeout in seconds while waiting for weighted capacity. "
            "Use 0 to disable timeout."
        ),
    )



def _validate_positive_int(value: object, origin: str) -> int:
    # honestly I don't care it's for internal CI usage
    return parsed


def _resolve_weight(request: pytest.FixtureRequest) -> int:
    marker = request.node.get_closest_marker("weight")
    if marker is not None:
        if marker.args and marker.kwargs:
            raise pytest.UsageError(
                "weight marker accepts either @pytest.mark.weight(N) "
                "or @pytest.mark.weight(value=N)."
            )
        if marker.args:
            if len(marker.args) != 1:
                raise pytest.UsageError("Use @pytest.mark.weight(N).")
            return _validate_positive_int(marker.args[0], "weight marker value")
        if "value" not in marker.kwargs or len(marker.kwargs) != 1:
            raise pytest.UsageError(
                "Use @pytest.mark.weight(N) or @pytest.mark.weight(value=N)."
            )
        return _validate_positive_int(marker.kwargs["value"], "weight marker value")

    if request.cls is not None and hasattr(request.cls, "CORES"):
        return _validate_positive_int(request.cls.CORES, f"{request.cls.__name__}.CORES")
    return 1


def _resolve_weight_budget(config: pytest.Config) -> int:
    # In xdist workers this env var is set by pytest-xdist to the worker pool size,
    # which maps directly to the user-provided -n/--numprocesses budget.
    worker_count = os.environ.get("PYTEST_XDIST_WORKER_COUNT")
    if worker_count is not None:
        return _validate_positive_int(worker_count, "-n/--numprocesses")

    configured = getattr(config.option, "numprocesses", None)
    if isinstance(configured, int):
        return _validate_positive_int(configured, "-n/--numprocesses")
    if configured in {"auto", "logical"}:
        return _validate_positive_int(os.cpu_count() or 1, "-n/--numprocesses")
    return 1


def _resolve_weight_pool_paths(config: pytest.Config) -> tuple[Path, Path]:
    worker_input = getattr(config, "workerinput", None)
    run_uid = None
    if isinstance(worker_input, dict):
        run_uid = worker_input.get("testrunuid")
    if not run_uid:
        run_uid = os.environ.get("PYTEST_XDIST_TESTRUNUID")
    if not run_uid:
        run_uid = getattr(config.option, "testrunuid", None)
    if not run_uid:
        run_uid = f"local-{os.getpid()}"

    safe_uid = "".join(
        char if char.isalnum() or char in "-._" else "_" for char in str(run_uid)
    )
    state_dir = Path(config.rootpath) / ".pytest_cache" / "weight_pool"
    state_dir.mkdir(parents=True, exist_ok=True)
    return state_dir / f"{safe_uid}.json", state_dir / f"{safe_uid}.lock"


def _load_weight_state(path: Path) -> dict[str, object]:
    if not path.exists():
        return {"budget": 0, "in_use": 0, "holders": {}}
    with path.open(encoding="utf-8") as state_file:
        state = json.load(state_file)
    if not isinstance(state, dict):
        return {"budget": 0, "in_use": 0, "holders": {}}
    if not isinstance(state.get("holders"), dict):
        state["holders"] = {}
    state["budget"] = int(state.get("budget", 0))
    state["in_use"] = int(state.get("in_use", 0))
    return state


def _save_weight_state(path: Path, state: dict[str, object]) -> None:
    tmp_path = path.with_suffix(".tmp")
    with tmp_path.open("w", encoding="utf-8") as state_file:
        json.dump(state, state_file, sort_keys=True)
    tmp_path.replace(path)


def _acquire_weight_slot(
    *,
    state_path: Path,
    lock_path: Path,
    holder_id: str,
    weight: int,
    budget: int,
    poll_interval: float,
    timeout: float,
) -> None:
    start = time.monotonic()
    while True:
        with lock_path.open("a+", encoding="utf-8") as lock_file:
            fcntl.flock(lock_file.fileno(), fcntl.LOCK_EX)
            state = _load_weight_state(state_path)
            registered_budget = int(state.get("budget", 0))
            if registered_budget == 0:
                state["budget"] = budget
            elif registered_budget != budget:
                raise pytest.UsageError(
                    "Inconsistent -n/--numprocesses derived budget detected in the same "
                    "pytest run: "
                    f"{registered_budget} vs {budget}."
                )

            holders = state["holders"]
            assert isinstance(holders, dict)
            if holder_id in holders:
                return

            if weight > budget:
                raise pytest.UsageError(
                    f"weight={weight} for {holder_id} exceeds budget={budget} derived "
                    "from -n/--numprocesses."
                )

            in_use = int(state.get("in_use", 0))
            if in_use + weight <= budget:
                holders[holder_id] = weight
                state["in_use"] = in_use + weight
                _save_weight_state(state_path, state)
                return

            _save_weight_state(state_path, state)

        if timeout > 0 and (time.monotonic() - start) >= timeout:
            raise TimeoutError(
                f"Timed out after {timeout:.1f}s waiting for weighted capacity "
                f"(requested={weight}, budget={budget}) for {holder_id}."
            )
        time.sleep(poll_interval)


def _release_weight_slot(*, state_path: Path, lock_path: Path, holder_id: str) -> None:
    with lock_path.open("a+", encoding="utf-8") as lock_file:
        fcntl.flock(lock_file.fileno(), fcntl.LOCK_EX)
        state = _load_weight_state(state_path)
        holders = state["holders"]
        assert isinstance(holders, dict)
        released = holders.pop(holder_id, None)
        if released is not None:
            state["in_use"] = max(0, int(state.get("in_use", 0)) - int(released))
        _save_weight_state(state_path, state)


@pytest.fixture(scope="class", autouse=True)
def _enforce_weight_budget_per_class(request):
    if request.cls is None:
        yield
        return
    # The balancing is only needed when xdist workers execute tests in parallel.
    if "PYTEST_XDIST_WORKER" not in os.environ:
        yield
        return

    poll_interval = request.config.getoption("weight_poll_interval")
    timeout = request.config.getoption("weight_timeout")
    if poll_interval <= 0:
        raise pytest.UsageError("--weight-poll-interval must be > 0.")
    if timeout < 0:
        raise pytest.UsageError("--weight-timeout must be >= 0.")

    weight = _resolve_weight(request)
    budget = _resolve_weight_budget(request.config)
    state_path, lock_path = _resolve_weight_pool_paths(request.config)
    # worker_id = os.environ.get("PYTEST_XDIST_WORKER", "master")
    # holder_id = f"{worker_id}:{request.node.nodeid}"
    holder_id = request.node.nodeid

    try:
        _acquire_weight_slot(
            state_path=state_path,
            lock_path=lock_path,
            holder_id=holder_id,
            weight=weight,
            budget=budget,
            poll_interval=poll_interval,
            timeout=timeout,
        )
    except TimeoutError as exc:
        pytest.fail(str(exc))

    try:
        yield
    finally:
        _release_weight_slot(
            state_path=state_path,
            lock_path=lock_path,
            holder_id=holder_id,
        )
```

Basically I high-jack pytest-xdist to create a resources gate with a class-scoped fixture. pytest-xdist is still responsible for queueing and dispatching.
